capsules/extra/src/nmea.rs

```diff
@@ -0,0 +1,245 @@
+                            .ok();
+                    }
+                } else if app.operation == Operation::Write {
+                    app.operation = Operation::None;
+
+                    // Only a single application can write at a time
+                    if status.is_err() {
+                        kernel_data
+                            .schedule_upcall(0, (into_statuscode(Err(ErrorCode::FAIL)), 0, 0))
+                            .ok();
+                    } else {
+                        kernel_data
+                            .schedule_upcall(0, (into_statuscode(Ok(())), len, 0))
+                            .ok();
+                    }
+                }
+            });
+        }
+
+        self.buffer.replace(buffer);
+        self.operation.set(Operation::None);
+    }
+}
+
+impl SyscallDriver for Nmea<'_> {
+    fn command(
+        &self,
+        command_num: usize,
+        length: usize,
+        _: usize,
+        processid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            // check whether the driver exists!!
+            0 => CommandReturn::success(),
+
+            // Read a sentence
+            // This is virtulised across all applications, so multiple
+            // applications can read data.
+            // Read/write operations can not occur in parrallel though.
+            // A write can change GNSS settings so we need to ensure
+            // that completes before we try to read again.
+            1 => {
+                if self.operation.get() == Operation::Write {
+                    // A write is ongoing, we need to let that finish
+                    return CommandReturn::failure(ErrorCode::BUSY);
+                }
+
+                self.apps
+                    .enter(processid, |app, _| {
+                        app.operation = Operation::Read;
+                        self.operation.set(Operation::Read);
+
+                        // If the buffer is already in use we return success
+                        // and the app will be notified when the curren read
+                        // operation completes
+                        self.buffer.take().map_or(CommandReturn::success(), |buf| {
+                            match self.driver.read_sentence(buf) {
+                                Ok(()) => CommandReturn::success(),
+                                Err((e, buffer)) => {
+                                    self.buffer.replace(buffer);
+                                    app.operation = Operation::None;
+                                    CommandReturn::failure(e)
+                                }
+                            }
+                        })
+                    })
+                    .unwrap_or_else(|err| CommandReturn::failure(err.into()))
+            }
+
+            // Write sentence
+            // Only a single write operation can occur at a time.
```

Comment:
Capsules should either be restricted to one app (not ideal) or virtualized.

Reply:
Yep. That's what this does

Reply:
I think currently it does neither. Restricted to one app means exactly one process can write, not one process _at a time_ can have an outstanding write. Actually virtualized means that each process can at least one outstanding write at a time, and the driver will take care of sending them in some order.

Reply:
I can change it so that only a single app can write.

I want to avoid write virtulisation as I don't think that makes sense. If one application is mid writing a series of commands we don't want a different application writing something in the middle of that and messing everything up.

Write virtulisation I suspect will be complex as we need userspace to signal start and stop of transactions. For now it seems unnecessary